documentation/design-docs/ipc-protocol.md

```diff
@@ -0,0 +1,183 @@
+# Event Pipe
+
+## Overview
+
+> TODO: Write an overview of what this does
+
+## TODOs
+* formalize struct names
+* formalize error codes (Use HRESULT values?)
+
+## Inter-Process Communication Specification
+
+### Current
+
+```c++
+enum class DiagnosticMessageType : uint32_t
+{
+    // EventPipe
+    StartEventPipeTracing = 1024, // To file
+    StopEventPipeTracing,
+    CollectEventPipeTracing, // To IPC
+};
+
+struct MessageHeader
+{
+    DiagnosticMessageType RequestType;
+    uint32_t Pid;
+};
+```
+
+```
+runtime <- client : MessageHeader { CollectEventPipeTracing }
+    error? -> no response
+runtime -> client : session ID 
+runtime -> client : event stream
+
+...
+
+runtime <- client : stop command
+runtime -> client : session id and stops
+```
+
+### Proposed
+
+Connection:
+```
+runtime <- client : handshake (set length byte array that represents version)
+runtime -> client : sends handshake back as ACK
+runtime <- client : packet containing start code
+runtime -> client : ACK + potential payload
+repeat till closed
+```
+
+### Transport considerations
+
+* All data will be sent in the endianess of the machine
+* All data will be sent in a packet containing a header of information concerning the contents
+
+### Handshake
+
+#### Requirements:
+* simple - easy to code
+* fast - no complex structures/serialization
+* extensible - leaves room to change resulting behavior
+
+The client will initiate the connection by sending a byte array containing the ASCII codes for the version string of the protocol it wishes to communicate with.
+
+> There will only be one for now, but it will give us the ability to extend this and potentially have _other_ protocols we can speak if we'd like.
+
+```c
+// "EVENTPIPE_V1"
+char[15] handshake = { 0x00, 0x00, 0x00, 0x45, 0x56, 0x45, 0x4e, 0x54, 0x50, 0x49, 0x50, 0x45, 0x5f, 0x56, 0x31 };
+```
+
+The `Accept` function in the runtime will immediately try to read 15 bytes of data and then use that determine what protocol we are speaking.  Once it has been determined, the runtime will respond with that handshake.
+
+#### Errors
+* runtime: 
+  * read failure -> end connection; no response
+  * unknown protocol -> ERROR response // TODO: flesh out errors
```

Comment:
I think it would be helpful to have an explicit error response before the dropped connection to help with versioning.  It shouldn't be any more complicated than simply dropping the connection and gives us more to work with on the consumer side.

Reply:
Go for it : )